### PR TITLE
[GPU] Optimize out Reshape and Reorder for outer axis padding

### PR DESCRIPTION
**Pull request by byungilm**
### Details:
 - Optimize batch axis eltwise

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Comment by isanghao** [comment]
Performance was good on dGPU: ./check_performance.py --this ../log/pr/2023-07-04--00-13-PR18163.report --ref ../log/2023-07-04--09-28.report

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h [hunk 1]
```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
...
+            for (size_t i = 0 ; i < input_layout.get_spatial_rank() ; i++) {
+                first_element_offset *= input_layout.spatial(i);
+            }
+            first_element_offset *= input_layout.feature();
+
+            size_t inner_size = 1;
+            for (size_t i = 0 ; i < output_layout.get_spatial_rank() ; i++) {
+                inner_size *= output_layout.spatial(i);
+            }
+            inner_size *= output_layout.feature();
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/include/reshape_inst.h:80]: this code assumes that the reshape may change batch size as well. But do we have such case in swin/twin/vit? If not, I think we can simplify the code here..

**Review by isanghao — CHANGES_REQUESTED** [review]

**Comment by hyunback** [comment]
@byungilm  It seems PR applied offset in weight, bias now, but I can't find onednn convolution, is it correct?

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h [hunk 2]
```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
...
+
+            auto new_batch_pad = first_element_offset / inner_size;
+            this->set_output_padding(cldnn::padding({static_cast<int32_t>(new_batch_pad), 0, 0, 0}, {0, 0, 0, 0}));
+
+            return;
```
Comment (hyunback) [inline src/plugins/intel_gpu/src/graph/include/reshape_inst.h:85]: nit: unnecessary

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h [hunk 3]
```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
...
+        }
+
+        return;
```
Comment (hyunback) [inline src/plugins/intel_gpu/src/graph/include/reshape_inst.h:88]: nit: unnecessary

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h [hunk 4]
```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
...
+    }
+
     bool is_in_place() const {
         if (this->is_output() || this->has_fused_primitives())
             return false;
-        return (!this->get_output_layout().data_padding && !input().get_output_layout(false).data_padding);
+
+        if (has_padding()) {
+            if (!this->is_dynamic() && has_outer_padding_offset() &&
+                this->get_users().front()->get_preferred_impl_type() == impl_types::onednn)
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:53]: I don't think so, My understanding is correct, all changes in reshape_inst.h are for reshape user is oneDNN only. Correct? If true, reshape_inst.h is not a good choice for specific onednn case.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 5]
```diff
@@ -554,6 +571,13 @@ void prepare_buffer_fusing::run(program& p) {
 
         program_helpers::do_for_types<reshape>(*node, [](reshape_node& node) {
             node.get_output_layout();
+
+            // Optimizing at prepare_buffer_fusing could propagate a padded input of an input nodes to Reshape.
+            // Reshape can be optimized out when only an outer axis(batch) has padding.
+            // For this case , it should re-calculate output padding size.
```
Comment (hyunback) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:577]: Below code is only for oneDNN reshape users, If true please add comment and check logic onednn, 
Maybe the better place than reshape_inst.h

File: src/plugins/intel_gpu/src/plugin/ops/matmul.cpp [hunk 6]
```diff
@@ -195,15 +195,17 @@ static void CreateMatMulOp(Program& p, const std::shared_ptr<ngraph::op::v0::Mat
...
-            auto is_u8_i8 = (type == ov::element::Type_t::i8 || type == ov::element::Type_t::u8);
             bool in0_large = std::all_of(shapes[0].rbegin(), shapes[0].rbegin() + 2,
                                         [] (const ngraph::Dimension& dim) { return dim.is_static() && dim.get_length() >= 64; });
             bool in1_large = std::all_of(shapes[1].rbegin(), shapes[1].rbegin() + 2,
                                         [] (const ngraph::Dimension& dim) { return dim.is_static() && dim.get_length() >= 64; });
+            // Optimized for clDNN
+            auto is_u8_i8 = (type == ov::element::Type_t::i8 || type == ov::element::Type_t::u8);
             bool in0_very_large = tensor_from_dims(shapes[0].to_shape()).count() > 100000;
             bool in1_very_large = tensor_from_dims(shapes[0].to_shape()).count() > 100000;
+            bool can_transpose_inputs = (in0_very_large || in1_very_large) && !is_u8_i8 && !p.get_engine().get_device_info().supports_immad;
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/plugin/ops/matmul.cpp:206]: Applied

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 7]
```diff
@@ -1302,13 +1302,9 @@ bool layout_optimizer::are_layouts_suitable_for_onednn(program_node& node) {
...
         bool no_batch_padding = true;
-        for (size_t i = 0; i < in_padding.lower_size().batch.size(); ++i) {
-            no_batch_padding &= (in_padding.lower_size().batch[i] == 0);
-        }
-        for (size_t i = 0; i < in_padding.upper_size().batch.size(); ++i) {
-            no_batch_padding &= (in_padding.upper_size().batch[i] == 0);
-        }
         for (size_t i = 0; i < out_padding.lower_size().batch.size(); ++i) {
             no_batch_padding &= (out_padding.lower_size().batch[i] == 0);
         }
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1310]: Yes. Thanks. I missed it.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 8]
```diff
@@ -523,14 +523,31 @@ void prepare_buffer_fusing::run(program& p) {
...
-                                out_pad.lower_size().spatial[1]},
+                                out_pad.lower_size().spatial[1],
+                                out_pad.lower_size().spatial[2]},
                                 {opt_upper_pad,
                                 out_pad.upper_size().feature[0],
                                 out_pad.upper_size().spatial[0],
-                                out_pad.upper_size().spatial[1]});
+                                out_pad.upper_size().spatial[1],
+                                out_pad.upper_size().spatial[2]});
+                    } else {
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:547]: Applied

File: src/plugins/intel_gpu/src/graph/impls/onednn/deconvolution_onednn.cpp [hunk 9]
```diff
@@ -36,12 +36,14 @@ struct deconvolution_onednn : typed_primitive_onednn_impl<deconvolution> {
 
         {
             auto weights = instance.weights_memory();
-            args.insert({DNNL_ARG_WEIGHTS, weights->get_onednn_memory(_pd.weights_desc(0))});
+            auto offset = onednn::get_offset(instance.get_input_layout(1), _pd.dnnl::primitive_desc_base::weights_desc(0));
+            args.insert({DNNL_ARG_WEIGHTS, weights->get_onednn_memory(_pd.weights_desc(0), offset)});
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/impls/onednn/deconvolution_onednn.cpp:40]: Added test-cases for conv and deconv.

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h [hunk 10]
```diff
@@ -23,10 +23,69 @@ struct typed_program_node<reshape> : public typed_program_node_base<reshape> {
...
+
+    bool has_outer_padding_offset() const {
+        if (!has_padding())
+            return false;
+
+        auto input_layout = input().get_output_layout(false);
+        auto input_pad = input_layout.data_padding;
+        bool spatial_pad = false;
+        for (size_t i = 0 ; i < input_layout.get_spatial_rank() ; i++)
+            spatial_pad |= (input_pad.lower_size().spatial[i] != 0);
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:38]: Thanks. Fixed.

[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:38]: Looks good. Applied.

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:53]: HB is right. It is currently only for oneDNN. I could be perf gain for clDNN as well. But I won't implement it in this PR.
And I agree HB's comment. I will find other place to set oneDNN specific condition.

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h (hunk 2 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:85]: Fixed.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp (hunk 5 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:577]: I agree. It is moved into can_reshape_be_optimized.

File: src/plugins/intel_gpu/src/graph/include/reshape_inst.h (hunk 1 above)
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/include/reshape_inst.h:80]: Yes, there is this case in swin.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 11]
```diff
@@ -186,10 +186,17 @@ void combine_bf_with_first_spatial_dim(cldnn::layout& l) {
...
     int64_t offset = 0;
+    auto b_padding = l.data_padding.lower_size().batch[0];
     auto f_padding = l.data_padding.lower_size().feature[0];
-    if (f_padding != 0) {
+    if (b_padding != 0) {
+        offset = b_padding;
+        for (size_t i = 0; i < l.get_spatial_rank(); ++i) {
+            offset *= l.spatial(i);
+        }
+        offset *= l.feature();
```
[the remark this answers is not in the document]
Comment (byungilm) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:198]: Thanks. I modified to use pitch size and added exception for multi blocked format.

**Review by hyunback — APPROVED** [review]

**Review by isanghao — APPROVED** [review]

**Pull request merged**